spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/CommitStore.scala

```diff
@@ -0,0 +1,153 @@
+ *  |   yes     |   no      | transient error (e.g. network hiccup)                           |
+ *  |   yes     |   yes     | physical conflict (allowed to rebase and retry)                 |
+ */
+class CommitFailedException(
+    retryable: Boolean, conflict: Boolean, message: String) extends Exception(message)
+
+/** Response container for [[CommitStore.commit]] API */
+case class CommitResponse(commit: Commit)
+
+/** A container class to inform the CommitStore about any changes in Protocol/Metadata */
+case class UpdatedActions(
+  commitInfo: CommitInfo,
+  newMetadata: Option[Metadata],
+  newProtocol: Option[Protocol])
+
+/**
+ * CommitStore is responsible for managing commits for a managed-commit delta table.
+ * 1. It provides API to commit a new version of the table. See [[CommitStore.commit]] API.
+ * 2. It makes sure that commits are backfilled if/when needed
+ * 3. It also tracks and returns unbackfilled commits. See [[CommitStore.getCommits]] API.
+ */
+trait CommitStore {
+  /**
+   * API to commit the given set of `actions` to the table represented by given `tablePath` at the
+   * given `commitVersion`.
+   * @return CommitResponse which contains the file status of the commit file. If the commit is
+   *         already backfilled, then the fileStatus could be omitted from response and the client
+   *         could get the info by themselves.
+   */
+  def commit(
+    logStore: LogStore,
+    hadoopConf: Configuration,
+    tablePath: Path,
+    commitVersion: Long,
+    actions: Iterator[String],
+    updatedActions: UpdatedActions): CommitResponse
+
+  /**
+   * API to get the un-backfilled commits for the table represented by the given `tablePath`.
+   * Commits older than `startVersion`, or newer than `endVersion` (if given), are ignored. The
+   * returned commits are contiguous and in ascending version order.
+   * Note that the first version returned by this API may not be equal to the `startVersion`. This
+   * happens when few versions starting from `startVersion` are already backfilled and so
+   * CommitStore may have stopped tracking them.
+   *
+   * @return a sequence of [[Commit]] which are tracked by [[CommitStore]].
+   */
+  def getCommits(
+    tablePath: Path,
+    startVersion: Long,
+    endVersion: Option[Long] = None): Seq[Commit]
+}
+
+/** A builder interface for CommitStore */
+trait CommitStoreBuilder {
+
+  /** Name of the commit-store */
+  def name: String
+
+  /** Returns a commit store based on the given conf */
+  def build(conf: Map[String, String]): CommitStore
+}
+
+/** Factory to get the correct CommitStore for a table */
+object CommitStoreProvider {
+  // mapping from different commit-owner names to the corresponding [[CommitStoreBuilder]]s.
+  private val nameToBuilderMapping = mutable.Map.empty[String, CommitStoreBuilder]
+
+  /** Returns a [[CommitStore]] for the given `name` and `conf` */
+  def getCommitStore(name: String, conf: Map[String, String]): CommitStore = {
```

Comment:
We should probably add `synchronized` to this and other methods that access the build map? 
(only the "clear" method has it right now)

Reply:
Adding `syncronized` to registerBuilder. Do you see a need to syncronize on reader? i.e. getCommitStore method?

Reply:
The reader definitely needs `synchronized` as well. It's not physically safe to read a normal `Map` that could be modified concurrently by other threads, because internal data structures could be(come) inconsistent during the read.

Reply:
(granted, we don't expect any concurrency in the common case... but `synchronized` is ~free in single-thread env and saves us from obscure errors if multi-thread races ever happen)